src/auth/passkey-config.ts

```diff
@@ -0,0 +1,188 @@
+      );
+    }
+    if (!validator.isNonNullObject(passkeyConfigRequest)) {
+      throw new FirebaseAuthError(
+        AuthClientErrorCode.INVALID_ARGUMENT,
+        "'passkeyConfigRequest' must not be null.",
+      );
+    }
+    const validKeys = {
+      expectedOrigins: true,
+    };
+    // Check for unsupported top-level attributes.
+    for (const key in passkeyConfigRequest) {
+      if (!(key in validKeys)) {
+        throw new FirebaseAuthError(
+          AuthClientErrorCode.INVALID_ARGUMENT,
+          `'${key}' is not a valid PasskeyConfigRequest parameter.`,
+        );
+      }
+    }
+    if (!validator.isNonEmptyArray(passkeyConfigRequest.expectedOrigins)) {
+      throw new FirebaseAuthError(
+        AuthClientErrorCode.INVALID_ARGUMENT,
+        "'passkeyConfigRequest.expectedOrigins' must contain at least one item.",
+      );
+    }
+    for (const origin of passkeyConfigRequest.expectedOrigins) {
+      if (!validator.isNonEmptyString(origin)) {
+        throw new FirebaseAuthError(
+          AuthClientErrorCode.INVALID_ARGUMENT,
+          "'passkeyConfigRequest.expectedOrigins' cannot contain empty strings.",
+        );
+      }
+    }    
+  }
+
+  /**
+   * Build a server request for a Passkey Config object.
+   * @param isCreateRequest - A boolean stating if it's a create request.
+   * @param passkeyConfigRequest - Passkey config to be updated.
+   * @param rpId - (optional) Relying party ID for the request if it's a create request.
+   * @returns The equivalent server request.
+   * @throws FirebaseAuthError - If validation fails.
+   * 
+   * @internal
+   */
+  public static buildServerRequest(isCreateRequest: boolean, passkeyConfigRequest?: PasskeyConfigRequest, 
+    rpId?: string): PasskeyConfigClientRequest {
+    PasskeyConfig.validate(isCreateRequest, passkeyConfigRequest, rpId);
+    const request: PasskeyConfigClientRequest = {};
+    if (isCreateRequest && typeof rpId !== 'undefined') {
+      request.rpId = rpId;
+    }
+    if (typeof passkeyConfigRequest?.expectedOrigins !== 'undefined') {
+      request.expectedOrigins = passkeyConfigRequest.expectedOrigins;
+    }
+    return request;
+  }
+
+  /**
+   * The Passkey Config object constructor.
+   * @param response - The server-side response used to initialize the Passkey Config object.
+   * @constructor
+   * 
+   * @internal
+   */
+  constructor(response: PasskeyConfigServerResponse) {
+    if (typeof response.name !== 'undefined') {
+      this.name = response.name;
+    }
+    if (typeof response.rpId !== 'undefined') {
+      this.rpId = response.rpId;
+    }
+    if (typeof response.expectedOrigins !== 'undefined') {
+      this.expectedOrigins = response.expectedOrigins;
+    }
+  }
+
+  /**
+   * Returns a JSON-serializable representation of this object.
+   * @returns A JSON-serializable representation of this object.
+   */
+  public toJSON(): object {
+    const json = {
+      name: deepCopy(this.name),
```

Comment:
for string value, do we need to use deepCopy?

Reply:
thanks for catching this!